source/cl/source/extension/source/intel_required_subgroup_size.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) Codeplay Software Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License") with LLVM
+// Exceptions; you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/LICENSE.txt
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <CL/cl_ext.h>
+#include <cl/device.h>
+#include <cl/kernel.h>
+#include <extension/intel_required_subgroup_size.h>
+
+extension::intel_required_subgroup_size::intel_required_subgroup_size()
+    : extension("cl_intel_required_subgroup_size",
+#ifdef OCL_EXTENSION_cl_intel_required_subgroup_size
+                usage_category::DEVICE
+#else
+                usage_category::DISABLED
+#endif
+                    CA_CL_EXT_VERSION(1, 0, 0)) {
+}
+
+cl_int extension::intel_required_subgroup_size::GetDeviceInfo(
+    cl_device_id device, cl_device_info param_name, size_t param_value_size,
+    void *param_value, size_t *param_value_size_ret) const {
+  // Only report extension if the extension wasn't disabled in cmake.
+  if (usage_category::DISABLED == usage) {
+    return CL_INVALID_VALUE;
+  }
+#if !defined(CL_VERSION_3_0)
+  return CL_INVALID_VALUE;
+#endif
+
+  if (CL_DEVICE_SUB_GROUP_SIZES_INTEL == param_name) {
+    // TODO: Query the device for the sizes it reports.
+    uint64_t num_sizes = 0;
+    const size_t param_size = num_sizes * sizeof(size_t);
```

Comment:
won't this be zero (num_sizes = 0)?

Reply:
Yeah, This was more so that it'd be easier to flesh out properly, since all the code's there already. If it's confusing I can remove it

Reply:
Good idea

Reply:
Though it's not technically "failing" so I won't mention that. If the user asks for how many sub-groups a device supports, we'll return 0, which is valid.